### ISSUE TITLE
POST /swarm/join as per API doc example never returns

### ISSUE THREAD

**Issue by jfbibeau** [moby/moby#24094]
**Output of `docker version`:**

```
Client:
 Version:      1.12.0-rc2
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   906eacd
 Built:        
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.0-rc2
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   906eacd
 Built:        
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 1
Server Version: 1.12.0-rc2
Storage Driver: devicemapper
 Pool Name: docker-253:1-167772289-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 468.8 MB
 Data Space Total: 107.4 GB
 Data Space Available: 41.24 GB
 Metadata Space Used: 1.208 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.146 GB
 Thin Pool Minimum Free Space: 10.74 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Either use `--storage-opt dm.thinpooldev` or use `--storage-opt dm.no_warn_on_loop_devices=true` to suppress this warning.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.107-RHEL7 (2015-10-14)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: overlay bridge null host
Swarm: active
 NodeID: 8bazscqlazvfeesb0ix0c5tmn
 IsManager: Yes
 Managers: 1
 Nodes: 1
 CACertHash: sha256:fa22cceba1f228fdac9d3954e02244779870f9193664be62d9200c8858247bb8
Runtimes: default
Default Runtime: default
Security Options: seccomp
Kernel Version: 3.10.0-327.el7.x86_64
Operating System: Red Hat Enterprise Linux Server 7.2 (Maipo)
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 5.671 GiB
Name: jf-a.localdomain
ID: FWPJ:7UNB:OLBN:23Y6:JFNN:ARRB:7MW6:RWIC:RG3P:I36U:RU47:3KE6
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Insecure Registries:
 127.0.0.0/8

```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
RHEL 7.2 VM on OpenStack.

**Steps to reproduce the issue:**
1. On Node A, initialize a swarm using the REST API example from the docs:

```
POST /swarm/init HTTP/1.1
Content-Type: application/json

{
  "ListenAddr": "0.0.0.0:4500",
  "ForceNewCluster": false,
  "Spec": {
    "AcceptancePolicy": {
      "Policies": [
        {
          "Role": "MANAGER",
          "Autoaccept": false
        },
        {
          "Role": "WORKER",
          "Autoaccept": true
        }
      ]
    },
    "Orchestration": {},
    "Raft": {},
    "Dispatcher": {},
    "CAConfig": {}
  }
}
```
1. On Node B, try to join using the REST API example from the docs, where x.x.x.x is the (OpenStack) floating IP address from node A:

```
POST /swarm/join HTTP/1.1
Content-Type: application/json

{
  "ListenAddr": "0.0.0.0:4500",
  "RemoteAddr": "x.x.x.x:4500",
  "Secret": "",
  "CACertHash": "",
  "Manager": false
}
```

First request always returns status code 500 (which is not a valid return code for this request, according to the docs):

```
{
  "message": "at least 1 RemoteAddr is required to join"
}
```

Even though I clearly have a `RemoteAddr` parameter. 

If I fire the exact same request a 2nd time, the request hangs forever.

**Describe the results you received:**
REST request hangs forever. Docker client commands such as `docker info` and `docker node ls` hang forever.

**Describe the results you expected:**
Node should join the swarm. 

**Additional information you deem important (e.g. issue happens only occasionally):**
100% reproducible. Running `docker swarm join x.x.x.x:4500` on the node itself works fine.

**Comment by vdemeester** [comment]
@jfbibeau `RemoteAddrs` is an *array*, so should work better with `["x.x.x.x:4500"]`.

```POST /swarm/join HTTP/1.1
Content-Type: application/json

{
  "ListenAddr": "0.0.0.0:4500",
  "RemoteAddr": ["x.x.x.x:4500"],
  "Secret": "",
  "CACertHash": "",
  "Manager": false
}
```

I'll update the reference API documentation.

**Comment by cpuguy83** [comment]
But why would it hang just because the field is wrong?

**Comment by jfbibeau** [comment]
@vdemeester I am still getting the same issue by sending an array. First request returns "{ "message": "at least 1 RemoteAddr is required to join" }", second request hangs the docker daemon.

**Comment by vdemeester** [comment]
Dang it, I made a typo in my comment :scream:… it's `RemoteAddrs` :sweat_smile:.
But as @cpuguy83, that does not explains why it hangs the 2nd time though... :thought_balloon:

**Comment by vdemeester** [comment]
Re-opening for the "hanging" part of the issue

**Comment by jfbibeau** [comment]
Wow, there's something really strange about the /join API. I wonder if it has something to do with the IP address I'm passing in to the join API (and the fact that it's an OpenStack floating IP maybe?).

When I use the correct form of the API (`RemoteAddrs` + array):
```
{
  "ListenAddr": "0.0.0.0:4500",
  "RemoteAddrs": ["x.x.x.x:4500"],
  "Secret": "",
  "CACertHash": "",
  "Manager": false
}
```

The API request goes through fine, however I get this as the response:
```
{
  "message": "Timeout was reached before node was joined. Attempt to join the cluster will continue in the background. Use \"docker info\" command to see the current Swarm status of your node."
}
```

The docker node never ends up joining the swarm. If I try to restart the docker service, now docker completely refuses to come up and fails with a GRPC error; I'm not sure how to recover from this at all as I can no longer start docker:

```
Jun 29 14:56:48 jf-a.localdomain docker[2660]: time="2016-06-29T14:56:48.359514758-04:00" level=info msg="Loading containers: start."
Jun 29 14:56:48 jf-a.localdomain docker[2660]: .time="2016-06-29T14:56:48.377843484-04:00" level=info msg="Firewalld running: false"
Jun 29 14:56:48 jf-a.localdomain docker[2660]: time="2016-06-29T14:56:48.477096302-04:00" level=info msg="Default bridge (docker0) is assigned with an IP address 172.17.0.0/16. Daemon option --bip can be used to set a preferred IP address"
Jun 29 14:56:48 jf-a.localdomain docker[2660]: time="2016-06-29T14:56:48.517797905-04:00" level=info msg="Loading containers: done."
Jun 29 14:56:57 jf-a.localdomain docker[2660]: time="2016-06-29T14:56:57.135323055-04:00" level=error msg="cluster exited with error: can't create raft node: rpc error: code = 2 desc = grpc: the client connection is closing"
Jun 29 14:56:57 jf-a.localdomain docker[2660]: time="2016-06-29T14:56:57.135404057-04:00" level=fatal msg="Error creating cluster component: swarm component could not be started"
Jun 29 14:56:57 jf-a.localdomain systemd[1]: docker.service: main process exited, code=exited, status=1/FAILURE
Jun 29 14:56:57 jf-a.localdomain systemd[1]: Failed to start Docker Application Container Engine.
Jun 29 14:56:57 jf-a.localdomain systemd[1]: Unit docker.service entered failed state.
Jun 29 14:56:57 jf-a.localdomain systemd[1]: docker.service failed.
```

**Comment by jfbibeau** [comment]
Ok! I managed to recover the system, I had to blow away /var/lib/docker (presumably the raft folder in it did it) to have docker come back up at all.

Also found the root cause, if I give the `RemoteAddrs` parameter the floating (public) IP of the docker master swarm node, that's what completely hoses the docker daemon.

If I give the internal IP of the VM as the `RemoteAddrs`, the API works properly. 

I'm not sure if this is helping you guys at all :O But there's definitely something funny going on in this area. Possibly 2 bugs:
1. one related to the API hanging when `RemoteAddr` is passed instead of `RemoteAddrs`
2. one related to the docker daemon not joining a cluster, and subsequently not starting up again, when a floating IP is passed as the `RemoteAddrs` parameter.

Anything else I can do to help, let me know.
